File: storage/data_transpiled/at_coder/abc051/D/4806449.java
public static void warshallFloyd ( int [ ] [ ] dist , int n ) {
  final Scanner input = new Scanner ( System . in ) ;
  final double inf = Double . MAX_VALUE ;
  /* WarshallFloyd the dist */
  for ( int k = 0 ;
  k < n ;
  k ++ ) {
    for ( int i = 0 ;
    i < n ;
    i ++ ) {
      for ( int j = 0 ;
      j < n ;
      j ++ ) {
        if ( i == j ) {
          dist [ i ] [ j ] = 0 ;
        }
        else if ( dist [ i ] [ j ] > dist [ i ] [ k ] + dist [ k ] [ j ] ) {
          dist [ i ] [ j ] = dist [ i ] [ k ] + dist [ k ] [ j ] ;
        }
      }
    }
  }
  final int N = Integer . parseInt ( input . nextLine ( ) ) ;
  final int M = Integer . parseInt ( input . nextLine ( ) ) ;
  final List < Edge > edges = new ArrayList < > ( ) ;
  final int [ ] [ ] ad = new int [ N ] [ N ] ;
  for ( int a = 0 ;
  a < M ;
  a ++ ) {
    for ( int b = 0 ;
    b < N ;
    b ++ ) {
      for ( int s = 0 ;
      s < N ;
      s ++ ) {
        if ( ad [ s ] [ a ] + c == ad [ s ] [ b ] ) {
          ans -- ;
          break ;
        }
      }
    }
  }
  System . out . println ( ans ) ;
}

File: storage/data_original/at_coder/abc051/D/4806449.py
import sys
input = sys.stdin.readline

inf = float('inf')

# dist[i][j]: i??j?????,  n:???
def warshall_floyd(dist,n):
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if i == j:
                    dist[i][j] = 0
                elif dist[i][j] > dist[i][k] + dist[k][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

N,M = map(int,input().split())
edges = []
ad = [[inf]*N for _ in range(N)]
for _ in range(M):
    a,b,c = map(int,input().split())
    a -= 1;  b -= 1
    edges.append((a,b,c))
    ad[a][b] = c
    ad[b][a] = c

warshall_floyd(ad,N)
ans = M

for a,b,c in edges:
    for s in range(N):
        if ad[s][a] + c == ad[s][b]:
            ans -= 1
            break

print(ans)